flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -882,6 +878,47 @@ else if (transitionState(current, ExecutionState.FAILED, t)) {
 		}
 	}
 
+	private void releaseNetworkResources() {
+		LOG.debug("Release task {} network resources (state: {}).", taskNameWithSubtask, getExecutionState());
+
+		for (ResultPartition partition : producedPartitions) {
+			taskEventDispatcher.unregisterPartition(partition.getPartitionId());
+		}
+
+		closeOrFailNetworkResources(isCanceledOrFailed(), getFailureCause());
+	}
+
+	/**
+	 * There are two scenarios to release the network resources. One is from {@link TaskCanceler} to early
+	 * release partitions and gates. Another is from task thread during task exiting.
+	 *
+	 * @param isFailed true if the task has failed.
+	 * @param cause the exception that caused the task to fail, or null, if the task has not failed.
+	 */
+	private void closeOrFailNetworkResources(boolean isFailed, @Nullable Throwable cause) {
```

Comment:
Thanks for de duplicating this logic already ( :) ), but you could also go one step further and maybe extract this logic to something like `TaskCloser` class and de-duplicate/re-use it in `TaskCanceller` as well?

Reply:
The introduced `TaskCloser` seems better, agree with it.

Reply:
I would suggest to pass then `Runnable closeNetworkResources` to `TaskCanceler` constructor because it is still task concern to manage partitions/gates and we can avoid exploding methods with parameters. `TaskCanceler` does not really closes the whole task, just `interrupt`s network resources as well.

Also, just thought, that `TaskCanceler` actually does not need `isFailed` case.
@zhijiangW maybe, we could simplify `closeOrFailNetworkResources` and do only closing there. If we remove `close` from `partition.fail`, we could move conditional `partition.fail` to loop in `releaseNetworkResources`.

Reply:
Oh... ok, sorry by doing quick review I missed one that you have already deduplicated `TaskCanceler` code 😳 I thought that those are two separate issues that I'm commenting on, but clearly this is just one thing, so responding to both threads here in single comment:

Yes, passing `invokable` is also not the pretties thing, but it looks like this issue is beside the scope of this PR, right?

I haven't thought it through but passing `Runnable closeNetworkResources` seems fine to me as well. It has a drawback of being more or less the same thing, but with more vague name/type in an exchange of less overhead code. With simple class `TaskCloser` we could better specify concurrency contracts (`@ThreadSafe`) etc, but I think I would be fine both way.